src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java

```diff
@@ -30,143 +38,272 @@
+                                                  Collection<FieldMapping> fieldMappings) {
+        Set<String> virtualFieldNames = fieldMappings.stream()
+            .map(m -> m.getSource().getName())
+            .collect(Collectors.toSet());
+        String newExpression = convertFieldNameLiteralsToExpressionNames(expression, request, virtualFieldNames);
 
-        convertFieldNameLiteralsToExpressionNames(request, fieldMappings);
-        fieldMappings.forEach(fieldMapping -> applyKeyConditionToField(request, fieldMapping));
+        fieldMappings.forEach(fieldMapping -> mapFieldInConditionExpression(newExpression, request, fieldMapping));
+        return newExpression;
     }
 
     /**
      * Finds a virtual field name reference in the expression attribute names, finds the value in the right-hand side
      * operand in the primary expression or filter expression, gets its value in the expression attribute values,
      * applies the mapping, and sets the physical name of the field to that of the target field.
-     */
-    private void applyKeyConditionToField(RequestWrapper request, FieldMapping fieldMapping) {
-        applyKeyConditionToField(
-            request,
-            fieldMapping,
-            request.getPrimaryExpression(),
-            request.getFilterExpression());
-    }
-
-    /**
-     * For a given virtual-physical field mapping, maps field names and applies field value prefixing for
-     * tenant isolation.
      *
      * <p>Comments show expected variable values with a sample set of inputs.
      */
     @VisibleForTesting
-    void applyKeyConditionToField(RequestWrapper request,
-                                  FieldMapping fieldMapping,
-                                  String primaryExpression, // "#field1 = :value"
-                                  String filterExpression) {
-        if (primaryExpression != null) {
-            String virtualAttrName = fieldMapping.getSource().getName(); // "virtualHk"
-            Map<String, String> expressionAttrNames = request.getExpressionAttributeNames(); // "#field1" -> "virtualHk"
-            Optional<String> keyFieldName = expressionAttrNames != null ? expressionAttrNames.entrySet().stream()
-                .filter(entry -> entry.getValue().equals(virtualAttrName)).map(Entry::getKey).findAny()
-                : Optional.empty(); // Optional[#field1]
-            if (keyFieldName.isPresent() && !keyFieldName.get().equals(NAME_PLACEHOLDER)) {
+    void mapFieldInConditionExpression(String expression, // "#field1 = :value"
+                                       RequestWrapper request,
+                                       FieldMapping fieldMapping) {
+        if (expression != null) {
+            String virtualField = fieldMapping.getSource().getName();
+            Set<String> fieldPlaceholders = getPlaceholdersForVirtualField(virtualField, request); // {#field1}
+
+            for (String fieldPlaceholder : fieldPlaceholders) {
```

Comment:
The diff here is weird, but given a placeholder and a field mapping, the logic is basically the same as before